File: datastruct/tree/binary-search-tree/src/BSTree.hs
{-
 BSTree.hs
 Copyright (C) 2010 Liu Xinyu (liuxinyu95@gmail.com)

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.

-}

module BSTree where

import Test.QuickCheck
import qualified Data.List as L -- for verification purpose
import Prelude hiding(min, max)

-- alternative to define tree Foldable:
--   {-# LANGUAGE DeriveFoldable #-}
--   This provide null, fold, elem automatically

data Tree a = Empty
            | Node (Tree a) a (Tree a) deriving (Show, Eq)

leaf a = Node Empty a Empty

left (Node l _ _) = l
left _ = Empty

right (Node _ _ r) = r
right _ = Empty

key (Node _ k _) = Just k
key _ = Nothing

isEmpty Empty = True
isEmpty _ = False

foldt _ _ z Empty = z
foldt f g z (Node l k r) = g (foldt f g z l) (f k) (foldt f g z r)

-- in-order traverse
mapt _ Empty = Empty
mapt f (Node l x r)= Node (mapt f l) (f x) (mapt f r)

maptr :: (a -> b) -> Tree a -> Tree b
maptr f = foldt f Node Empty

-- plain fold, drop the tree structure, i.e. can't implement map
fold _ z Empty = z
fold f z (Node l k r) = fold f (k `f` (fold f z r)) l

member _ Empty = False
member x (Node l k r) | x == k = True
                     | x < k = member x l
                     | otherwise = member x r

-- lookupt :: (Eq a, Ord a) => a -> Tree (a, b) -> Maybe b
lookupt _ Empty = Nothing
lookupt x (Node l (k, v) r) | k == x = Just v
                            | x < k = lookupt x l
                            | otherwise = lookupt x r

min (Node Empty k _) = k
min (Node l _ _) = min l

max (Node _ k Empty) = k
max (Node _ _ r) = max r

insert x Empty = Node Empty x Empty
insert x (Node l k r) | x < k = Node (insert x l) k r
                      | otherwise = Node l k (insert x r)

-- Delete an element from a tree
--   if x has only one child: just splice x out
--   if x has two children: use min(right) to replace x
delete _ Empty = Empty
delete x (Node l k r) | x < k = Node (delete x l) k r
                      | x > k = Node l k (delete x r)
                      | otherwise = del l r
  where
    del Empty r = r
    del l Empty = l
    del l r = let k' = min r in Node l k' (delete k' r)

-- Traverse a part of tree inside a range [a, b]
mapR f a b t = map' t where
    map' Empty = Empty
    map' (Node l k r) | k < a = map' r
                      | a <= k && k <= b = Node (map' l) (f k) (map' r)
                      | k > b = map' l

fromList::(Ord a) => [a] -> Tree a
fromList = foldr insert Empty

toList Empty = []
toList (Node l k r) = toList l ++ [k] ++ toList r

-- tree sort
tsort :: (Eq a, Ord a) => [a] -> [a]
tsort = toList . fromList

preOrder Empty = []
preOrder (Node l k r) = k : preOrder l ++ preOrder r

postOrder Empty = []
postOrder (Node l k r) = postOrder l ++ postOrder r ++ [k]

-- Rebuild the binary tree from pre-order/in-order traverse list
rebuild [] _ = Empty
rebuild [c] _ = leaf c
rebuild (x:xs) ins = Node (rebuild prl inl) x (rebuild prr inr) where
  (inl, _:inr) = (takeWhile (/= x) ins, dropWhile (/=x) ins)
  (prl, prr) = splitAt (length inl) xs

-- test
prop_build :: (Ord a) => [a] -> Bool
prop_build xs = L.sort xs == tsort xs

prop_map :: (Ord a) => [a] -> Bool
prop_map xs = mapt id (fromList xs) == fromList xs

prop_map1 :: (Ord a) => [a] -> Bool
prop_map1 xs = maptr id (fromList xs) == fromList xs

prop_fold :: (Ord a, Num a) => [a] -> Bool
prop_fold xs = fold (+) 0 (fromList xs) == sum xs

prop_elem :: (Ord a) => [a] -> a -> Bool
prop_elem xs x = (x `elem` xs) == (x `member` (fromList xs))

prop_lookup :: (Ord a) => [a] -> a -> Bool
prop_lookup xs x = (lookupt x (fromList kvs)) == (lookup x kvs) where
  kvs = zip (L.nub xs) [1..]

prop_min :: (Ord a, Num a) => [a] -> Property
prop_min xs = not (null xs) ==> minimum xs == min (fromList xs)

prop_max :: (Ord a, Num a) => [a] -> Property
prop_max xs = not (null xs) ==> maximum xs == max (fromList xs)

prop_del :: (Ord a, Num a) => [a] -> a -> Bool
prop_del xs x = L.sort (L.delete x xs) == toList (delete x (fromList xs))

prop_mapR :: (Ord a, Num a) =>[a] -> a -> a -> Bool
prop_mapR xs a b = filter (\x-> a<= x && x <=b) (L.sort xs) ==
                   toList (mapR id a b (fromList xs))

prop_rebuild :: (Ord a, Num a) => [a] -> Bool
prop_rebuild xs = tr == rebuild prs ins where
  tr = fromList xs
  prs = preOrder tr
  ins = toList tr

testAll = do
  quickCheck (prop_build::[Int]->Bool)
  quickCheck (prop_map::[Int]->Bool)
  quickCheck (prop_map1::[Int]->Bool)
  quickCheck (prop_fold::[Int]->Bool)
  quickCheck (prop_elem::[Int]->Int->Bool)
  quickCheck (prop_lookup::[Int]->Int->Bool)
  quickCheck (prop_min::[Int]->Property)
  quickCheck (prop_max::[Int]->Property)
  quickCheck (prop_del::[Int]->Int->Bool)
  quickCheck (prop_mapR::[Int]->Int->Int->Bool)
  quickCheck (prop_rebuild::[Int]->Bool)

File: datastruct/tree/binary-search-tree/src/scala/BSTree.scala
/*
 * BSTree.scala
 * Copyright (C) 2018 Liu Xinyu (liuxinyu95@gmail.com)
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * 
 */
import scala.util.Random                //for verification purpose
import scala.language.postfixOps

object BSTree {

  sealed trait Tree[+A]
  case object Empty extends Tree[Nothing]
  case class Node[A] (left: Tree[A], key: A, right: Tree[A]) extends Tree[A]

  // helpers
  def leaf[A] (x: A) : Tree[A] = Node(Empty, x, Empty)

  def isEmpty[A] (tr: Tree[A]): Boolean =
    tr match {
      case Empty => true
      case _ => false
    }

  // in-order traverse
  def map [A, B] (f: A => B, tr: Tree[A]): Tree[B] =
    tr match {
      case Empty => Empty
      case Node(left, x, right) => Node(map(f, left), f(x), map(f, right))
    }

  def lookup[A <% Ordered[A]] (tr: Tree[A], x: A): Tree[A] =
    tr match {
      case Empty => Empty
      case Node(left, y, right) =>
        if (x == y) tr
        else if (x < y) lookup(left, x)
        else lookup(right, x)
    }

  def min[A] (tr: Tree[A]): A =
    tr match {
      case Node(Empty, x, _) => x
      case Node(left, _, _) => min(left)
    }

  def max[A] (tr: Tree[A]): A =
    tr match {
      case Node(_, x, Empty) => x
      case Node(_, _, right) => max(right)
    }

  def insert[A <% Ordered[A]] (tr: Tree[A], x: A): Tree[A] =
    tr match {
      case Empty => Node(Empty, x, Empty)
      case Node(left, y, right) =>
        if (x < y) Node(insert(left, x), y, right)
        else Node(left, y, insert(right, x))
    }

  def delete[A <% Ordered[A]] (tr: Tree[A], x: A): Tree[A] =
    tr match {
      case Empty => Empty
      case Node(left, y, right) =>
        if (x < y) Node(delete(left, x), y, right)
        else if (y < x) Node(left, y, delete(right, x))
        else if (isEmpty(left)) right
        else if (isEmpty(right)) left
        else {
          val z = min(right)
          Node(left, z, delete(right, z))
        }
    }

  def fromList [A <% Ordered[A]] (xs: Seq[A]) : Tree[A] = ((Empty: Tree[A]) /: xs) (insert)

  def toList[A] (tr: Tree[A]) : List[A] =
    tr match {
      case Empty => List()
      case Node(left, x, right) => toList(left) ::: (x :: toList(right))
    }

  val N = 100
  def genList(r: Random) = r.shuffle(0 to N - 1).take(r.nextInt(N))

  def testBuild(xs: Seq[Int]) = {
    assert(toList(fromList(xs)) == xs.sortWith(_ < _), println("violate build invariant"))
  }

  def testLookup(xs: Seq[Int], x: Int) = {
    lookup(fromList(xs), x) match {
      case Empty => assert(!xs.contains(x), println(s"$x exists"))
      case Node(_, y, _) => assert(x == y, println(s"given $x, found $y"))
    }
  }

  def testMinMax(xs: Seq[Int]) = {
    if (!xs.isEmpty) {
      val tr = fromList(xs)
      assert(xs.min == min(tr), println(s"failed to find the min in $xs"))
      assert(xs.max == max(tr), println(s"failed to find the max in $xs"))
    }
  }

  def testDelete(xs: Seq[Int]) = {
    if (!xs.isEmpty) {
      xs.foldLeft((xs.sorted, fromList(xs))) {
        (t, x) => {
          val ys = t._1
          val tr = t._2
          assert(ys == toList(tr), println(s"inconsist delete result: $ys"))
          (ys diff List(x), delete(tr, x))
        }
      }
    }
  }

  def test() = {
    val r = Random
    for (_ <- 1 to N) {
      val xs = genList(r)
      testBuild(xs)
      testLookup(xs, r.nextInt(N))
      testMinMax(xs)
      testDelete(xs)
    }
    println(s"$N tests passed");
  }
}
